sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -118,6 +118,7 @@ case class AdaptiveSparkPlanExec(
     val ensureRequirements =
       EnsureRequirements(requiredDistribution.isDefined, requiredDistribution)
     Seq(
+      CoalesceBucketsInJoin,
```

Comment:
shall we put it in `queryStageOptimizerRules`?

Reply:
rules in `queryStageOptimizerRules` are invoked less often which is more efficient. The rule `CoalesceBucketsInJoin` does not change plan partitioning and seems can be put in `queryStageOptimizerRules`

Reply:
Can we spend a bit of time understanding why? Then we can write a code comment to explain it and following developers won't try to move this rule to `queryStageOptimizerRules` ever.

Reply:
`CoalesceBucketsInJoin` should before `EnsureRequirements`.